Public/Src/Cache/ContentStore/Distributed/Utilities/GrpcFileCopier.cs

```diff
@@ -0,0 +1,92 @@
+using System.IO;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+using BuildXL.Cache.ContentStore.Hashing;
+using BuildXL.Cache.ContentStore.Interfaces.FileSystem;
+using BuildXL.Cache.ContentStore.Interfaces.Results;
+using BuildXL.Cache.ContentStore.Interfaces.Tracing;
+using BuildXL.Cache.ContentStore.Interfaces.Utils;
+using BuildXL.Cache.ContentStore.Service.Grpc;
+
+namespace BuildXL.Cache.ContentStore.Distributed.Utilities
+{
+    /// <summary>
+    /// File copier which operates over Grpc. <seealso cref="GrpcCopyClient"/>, <seealso cref="GrpcServerFactory"/>
+    /// </summary>
+    public class GrpcFileCopier : IAbsolutePathFileCopier
+    {
+        private Context _context;
+        private int _grpcPort;
+
+        /// <summary>
+        /// Constructor for <see cref="GrpcFileCopier"/>.
+        /// </summary>
+        public GrpcFileCopier(Context context, int grpcPort)
+        {
+            _context = context;
+            _grpcPort = grpcPort;
+        }
+
+        /// <inheritdoc />
+        public Task<FileExistenceResult> CheckFileExistsAsync(AbsolutePath path, TimeSpan timeout, CancellationToken cancellationToken)
+        {
+            // TODO: Implement!
+            throw new NotImplementedException();
+        }
+
+        /// <inheritdoc />
+        public async Task<CopyFileResult> CopyFileAsync(AbsolutePath sourcePath, AbsolutePath destinationPath, long contentSize, bool overwrite, CancellationToken cancellationToken)
+        {
+            // Extract host and contentHash from sourcePath
+            (string host, ContentHash contentHash) = ExtractHostHashFromAbsolutePath(sourcePath);
+
+            CopyFileResult copyFileResult = null;
+            // Contact hard-coded port on source
+            using (var client = GrpcCopyClient.Create(System.Net.IPAddress.Any.ToString(), 7089))
+            {
+                copyFileResult = await client.CopyFileAsync(_context, contentHash, destinationPath, contentSize);
+            }
+
+            return copyFileResult;
+        }
+
+        private (string host, ContentHash contentHash) ExtractHostHashFromAbsolutePath(AbsolutePath sourcePath)
+        {
+            Contract.Assert(sourcePath.IsUnc);
+
+            // TODO: Keep the segments in the AbsolutePath object?
+            // TODO: Indexable structure?
+            var segments = sourcePath.GetSegments();
+            Contract.Assert(segments.Count >= 4);
+
+            var host = segments.First();
+            var hashLiteral = segments.Last();
+            if (hashLiteral.EndsWith(GrpcDistributedPathTransformer.BlobFileExtension))
+            {
+                hashLiteral = hashLiteral.Substring(0, hashLiteral.Length - GrpcDistributedPathTransformer.BlobFileExtension.Length);
+            }
+            var hashTypeLiteral = segments.ElementAt(segments.Count - 1 - 2);
+
+            if (!Enum.TryParse<HashType>(hashTypeLiteral, out HashType hashType))

```

Comment:
>hashType [](start = 71, length = 8)

Should we check that the length of the hash matches the length expected based on the type?

Reply:
This is validated by the `ContentHash` constructor

---
In reply to: [271394465](https://github.com/Microsoft/BuildXL/pull/66#discussion_r271394465) [](ancestors = 271394465)

Reply:
Thanks!

---
In reply to: [271469202](https://github.com/Microsoft/BuildXL/pull/66#discussion_r271469202) [](ancestors = 271469202,271394465)